str/qc/qc_annotator.py

```diff
@@ -0,0 +1,223 @@
+    aggregated_table = exploded_table.group_by(exploded_table.REPID).aggregate(
+        allele_array_counts=hl.agg.counter(exploded_table.alleles_rep_lengths)
+    )
+
+    # finds the mode allele at each locus
+    max_key_expr = hl.bind(
+        lambda d: hl.bind(
+            lambda kv: kv[0], hl.sorted(d.items(), key=lambda kv: -kv[1])[0]
+        ),
+        aggregated_table.allele_array_counts,
+    )
+    aggregated_table = aggregated_table.annotate(mode_allele=max_key_expr)
+
+    # annotate the mode allele and dictionary of frequency of each allele into mt
+    mt = mt.annotate_rows(aggregated_info=aggregated_table[mt.REPID])
+    # number of distinct alleles identified at each locus
+    mt = mt.annotate_rows(
+        num_alleles=hl.len(mt.aggregated_info.allele_array_counts.values())
+    )
+
+    # the difference between each allele and the mode allele
+    mt = mt.annotate_entries(
+        allele_1_minus_mode=mt.allele_1_rep_length - mt.aggregated_info.mode_allele
+    )
+    mt = mt.annotate_entries(
+        allele_2_minus_mode=mt.allele_2_rep_length - mt.aggregated_info.mode_allele
+    )
+
+    # annotate the sum of alleles that are not the mode allele
+    mt = mt.annotate_entries(
+        allele_1_is_non_mode=hl.if_else(mt.allele_1_minus_mode != 0, True, False)
+    )
+    mt = mt.annotate_entries(
+        allele_2_is_non_mode=hl.if_else(mt.allele_2_minus_mode != 0, True, False)
+    )
+    mt = mt.annotate_rows(
+        sum_allele_1_is_not_mode=hl.agg.sum(hl.cond(mt.allele_1_is_non_mode, 1, 0)),
+        sum_allele_2_is_not_mode=hl.agg.sum(hl.cond(mt.allele_2_is_non_mode, 1, 0)),
+    )
+    mt = mt.annotate_rows(
+        sum_alleles_is_not_mode=mt.sum_allele_1_is_not_mode
+        + mt.sum_allele_2_is_not_mode
+    )
+    mt = mt.drop('sum_allele_1_is_not_mode', 'sum_allele_2_is_not_mode')
+
+    # proportion of alleles that are not the mode allele
+    mt = mt.annotate_rows(
+        prop_alleles_is_not_mode=mt.sum_alleles_is_not_mode
+        / hl.sum(mt.aggregated_info.allele_array_counts.values())
+    )
+
+    # add in Gymrek binomial HWEP implementation (Hail Query's HWEP implementation only works for biallelic loci)
+    # https://github.com/gymrek-lab/TRTools/blob/master/trtools/utils/utils.py#L325
+    mt = mt.annotate_rows(n_hom=mt.variant_qc.n_called - mt.variant_qc.n_het)
+    mt = mt.annotate_rows(exp_hom_frac=hl.sum(mt.variant_qc.AF**2))
+    mt = mt.annotate_rows(n_hom=hl.int32(mt.n_hom))
+    mt = mt.annotate_rows(n_called=hl.int32(mt.variant_qc.n_called))
+    mt = mt.annotate_rows(
+        binom_hwep=hl.binom_test(mt.n_hom, mt.n_called, mt.exp_hom_frac, 'two-sided')
+    )
+    mt = mt.drop('n_hom', 'n_called', 'exp_hom_frac')
+
+    # proportion of observed heterozygous calls per locus
+    mt = mt.annotate_rows(obs_het=mt.variant_qc.n_het / mt.variant_qc.n_called)
+
+    # average 'locus coverage' per locus
+    mt = mt.annotate_rows(variant_lc=hl.agg.sum(mt.LC) / mt.variant_qc.n_called)
```

Comment:
I don't think there's a functional benefit either way, but you can condense these calls way down by annotating multiple attributes in a single `.annotate_X` call

```
mt = mt.annotate_rows(
    x=y,
    foo=bar
)
```